snapshot/storage/metastore.go

```diff
@@ -0,0 +1,83 @@
+package storage
+
+import (
+	"context"
+
+	"github.com/docker/containerd/snapshot"
+)
+
+// MetaStore is used to store metadata related to a snapshot driver. The
+// MetaStore is intended to store metadata related to name, state and
+// parentage. Using the MetaStore is not required to implement a snapshot
+// driver but can be used to handle the persistence and transactional
+// complexities of a driver implementation.
+type MetaStore interface {
+	// Stat returns the snapshot stat Info directly from
+	// the metadata.
+	Stat(ctx context.Context, key string) (snapshot.Info, error)
+
+	// Walk iterates through all metadata for the stored
+	// snapshots and calls the provided function for each.
+	Walk(ctx context.Context, fn func(context.Context, snapshot.Info) error) error
+
+	// CreateActive creates a new active snapshot transaction referenced by
+	// the provided key. The provided opts carries the parent and readonly
+	// properties to store with the active snapshots. Additionally a create
+	// method can be provided in order to allow the caller perform an action
+	// with the active snapshots identifier before the metadata is
+	// committed. This callback action should be as short as possible to
+	// avoid blocking the transaction longer than necessary.
+	CreateActive(ctx context.Context, key string, opts CreateActiveOpts) error
+
+	// GetActive returns the metadata for the active snapshot transaction
+	// referenced by the given key.
+	// Set allParents to true to return the entire parent chain. By default
+	// only the immediate parent is included.
+	GetActive(ctx context.Context, key string, allParents bool) (Active, error)
+
+	// Remove removes a snapshot from the metastore. Cleanup is called
+	// during the transaction to allow the driver to make data unavailable
+	// before committing. The cleanup callback should be treated as the
+	// first part in a multi-phase commit, meaning data should be marked
+	// for removal but restorable if the database commit does not succeed.
+	Remove(ctx context.Context, key string, cleanup func(id string, k snapshot.Kind) error) error
+
+	// Commit renames the active snapshot transaction referenced by `key`
+	// as a committed snapshot referenced by `Name`. The resulting snapshot
+	// will be committed and readonly and the `key` reference will no longer
```

Comment:
This implementation makes `Commit` terminal (matching the current overlay implementation). It does not match the model described in the `Snapshotter` definition but based on the conversations we had, the `Snapshotter` needs to change to add new method which just does `Snapshot` as a non-terminal action and the model accordingly.